src/Microsoft.AspNet.Routing/Constraints/HttpMethodRouteConstraint.cs

```diff
@@ -0,0 +1,96 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using Microsoft.AspNet.Http;
+
+namespace Microsoft.AspNet.Routing.Constraints
+{
+    /// <summary>
+    /// Constrains the HTTP method of request or a route.
+    /// </summary>
+    public class HttpMethodConstraint : IRouteConstraint
+    {
+        /// <summary>
+        /// Creates a new <see cref="HttpMethodConstraint"/> that accepts the HTTP methods specified
+        /// by <paramref name="allowedMethods"/>.
+        /// </summary>
+        /// <param name="allowedMethods">The allowed HTTP methods.</param>
+        public HttpMethodConstraint(params string[] allowedMethods)
+        {
+            if (allowedMethods == null)
+            {
+                throw new ArgumentNullException(nameof(allowedMethods));
+            }
+
+            AllowedMethods = new List<string>(allowedMethods);
+        }
+
+        /// <summary>
+        /// Gets the HTTP methods allowed by the constraint.
+        /// </summary>
+        public IList<string> AllowedMethods { get; }
+
+        /// <inheritdoc />
+        public virtual bool Match(
+            HttpContext httpContext,
+            IRouter route,
+            string routeKey,
+            RouteValueDictionary values,
+            RouteDirection routeDirection)
+        {
+            if (httpContext == null)
+            {
+                throw new ArgumentNullException(nameof(httpContext));
+            }
+
+            if (route == null)
+            {
+                throw new ArgumentNullException(nameof(route));
+            }
+
+            if (routeKey == null)
+            {
+                throw new ArgumentNullException(nameof(routeKey));
+            }
+
+            if (values == null)
+            {
+                throw new ArgumentNullException(nameof(values));
+            }
+
+            switch (routeDirection)
+            {
+                case RouteDirection.IncomingRequest:
+                    return AllowedMethods.Contains(httpContext.Request.Method, StringComparer.OrdinalIgnoreCase);
+
+                case RouteDirection.UrlGeneration:
+                    // We need to see if the user specified the HTTP method explicitly.  Consider these two routes:
```

Comment:
Good, I'm glad we have these codes here! This was perhaps the main thing I was concerned about us forgetting to bring in.